### PR TITLE
Add Android back button

### PR DESCRIPTION
**Pull request by louismarie**
Related to #463

### CODE REVIEW

**Comment by Crash--** [comment]
👍

File: detox/src/devices/Device.js [hunk 1]
```diff
@@ -173,6 +173,10 @@ class Device {
     await this.deviceDriver.cleanup(this._deviceId, this._bundleId);
   }
 
+  async pushBackButton() {
```
Comment (rotemmiz) [inline detox/src/devices/Device.js]: `pressBack` might be a more familiar name.

File: detox/src/devices/IosDriver.js [hunk 2]
```diff
@@ -71,6 +71,10 @@ class IosDriver extends DeviceDriverBase {
   getPlatform() {
     return 'ios';
   }
+
```
Comment (rotemmiz) [inline detox/src/devices/IosDriver.js]: Add `log.warn` saying something like "pressBack is an android specific function, make sure you create an android specific test for this scenario"

File: detox/src/devices/AndroidDriver.js [hunk 3]
```diff
@@ -32,6 +32,10 @@ class AndroidDriver extends DeviceDriverBase {
     await this.adb.install(deviceId, this.getTestApkPath(binaryPath));
   }
 
+  async pushBackButton(deviceId) {
+    await this.adb.pushBackButton(deviceId, `input keyevent 4`);
```
Comment (rotemmiz) [inline detox/src/devices/AndroidDriver.js]: I think the  `input keyevent 4` param is redundant

File: detox/src/devices/android/ADB.js [hunk 4]
```diff
@@ -92,6 +92,10 @@ class ADB {
   async sleep(ms = 0) {
     return new Promise((resolve, reject) => setTimeout(resolve, ms));
   }
+
+  async pushBackButton(deviceId) {
```
Comment (rotemmiz) [inline detox/src/devices/android/ADB.js]: pressBack

**Comment by heuism** [comment]
Hi @rotemmiz , Sorry to disturb but if this one is okay to be merged? or is there anything blocking this from being merged? my current one need the back button implementation for Detox so I just want to ask. Thanks

**Comment by ksmahesh** [comment]
hey @louismarie are you planning on landing changes for pressBackButton anytime soon? Or has this been abandoned?

**Comment by louismarie** [comment]
All related @rotemmiz comments have been fixed in the last commit. To me all necessary changes have been made in the last commit so what should I do in order to merge this PR ?

**Comment by LeoNatan** [comment]
@rotemmiz Can this be merged?

**Comment by lalka-workco** [comment]
Will this merge go through? Looking at the Jenkins build shows that the detox-android-pr hasn't passed well before this PR. Would like to use this feature if it's ready.

**Comment by FibreFoX** [comment]
@louismarie there seems to be a merge-conflict :) as I would like this feature to be part of detox, can you check this? Would help a lot

**Comment by rotemmiz** [comment]
Allow me to begin with an apology, I'm not sure why we did not continue communicating regarding this issue. I am sorry. 

We've made a few changes around Android invocation, [`UIDevice.java`](https://developer.android.com/reference/android/support/test/uiautomator/UiDevice.html) is [fully exposed](https://github.com/wix/detox/blob/master/detox/src/android/espressoapi/UIDevice.js) through our invocation mechanism. 
I prefer implementing the back button using UIDevice, it is [synchronized](https://android.googlesource.com/platform/frameworks/testing/+/79693ed/uiautomator/library/src/com/android/uiautomator/core/UiDevice.java#185) and should be safer than adb. 

Usage example can be found here: https://github.com/wix/detox/blob/e0f4614fb9c6f78752cace9c577ab5811bd54a32/detox/src/devices/AndroidDriver.js#L142

**Comment by rotemmiz** [comment]
@louismarie, if you need any help, ping me here.

**Comment by louismarie** [comment]
OK @rotemmiz I did it ! I Hope these changes will be adopted :)

**Comment by rotemmiz** [comment]
Any idea why all iOS builds fail ? :(

File: detox/test/e2e/06.device.test.js [hunk 5]
```diff
@@ -54,4 +54,41 @@ describe('Device', () => {
     await device.shake();
     await expect(element(by.text('Shaken, not stirred'))).toBeVisible();
   });
+
+  describe('device orientation', () => {
```
Comment (rotemmiz) [inline detox/test/e2e/06.device.test.js]: We moved orientation tests outside this file, you can delete this part

File: detox/test/e2e/06.device.test.js [hunk 6]
```diff
@@ -54,4 +54,41 @@ describe('Device', () => {
...
+    it('OrientationPortrait', async() => {
+      // As default is portrait we need to set it otherwise
+      await device.setOrientation('landscape');
+      await device.setOrientation('portrait');
+
+      await expect(element(by.id('currentOrientation'))).toHaveText('Portrait');
+    });
+  });
+
+  describe('device button', () => {
```
Comment (rotemmiz) [inline detox/test/e2e/06.device.test.js]: I see why iOS builds fail, add `:android:` to the test name to make sure iOS tests skip it.
Also, change the test name to something more descriptive if possible.

File: detox/test/e2e/06.device.test.js (hunk 5 above)
Comment (louismarie) [reply detox/test/e2e/06.device.test.js]: done

File: detox/test/e2e/06.device.test.js (hunk 6 above)
Comment (louismarie) [reply detox/test/e2e/06.device.test.js]: done

File: docs/APIRef.DeviceObjectAPI.md [hunk 7]
```diff
@@ -247,6 +248,13 @@ if (device.getPlatform() === 'ios') {
   await expect(loopSwitch).toHaveValue('1');
 }
 ```
- 
+
+### `device.pressBack()`
+Simulate press back button (Android only).
```
Comment (rotemmiz) [inline docs/APIRef.DeviceObjectAPI.md]: Please add **Android Only** in the title (and link) instead of the description.

Comment (louismarie) [reply docs/APIRef.DeviceObjectAPI.md]: OK

**Comment by noomorph** [comment]
The pull request has been merged into master.

**Pull request closed without merging**